base/util/PKICertImport.bash

```diff
@@ -258,36 +498,320 @@ function PKICertImport() {
     # Remove the certificate from the NSS DB specified by $1. Errors are fatal;
     # uses exit instead of return.
     function _remove_cert() {
-        local database="$1"
+        local nickname="$1"
+        local return_no_exit="$2"
+
         local remove_args=("-D")
 
-        remove_args+=("-d" "$NSSDB_TYPE$database")
+        remove_args+=("-d" "$NSSDB_TYPE$NSSDB")
         if [ "x$NSSDB_PASSWORD" != "x" ]; then
             remove_args+=("-f" "$NSSDB_PASSWORD")
         fi
 
-        __v certutil "${remove_args[@]}" "-n" "$CERT_NICKNAME"
-        certutil "${remove_args[@]}" "-n" "$CERT_NICKNAME"
+        __v certutil "${remove_args[@]}" "-n" "$nickname"
+        certutil "${remove_args[@]}" "-n" "$nickname"
         local ret=$?
         if (( ret != 0 )); then
             __e "certutil returned non-zero result: $ret"
             __e "Unable to delete certificate!"
-            exit $?
+
+            if [ "x$return_no_exit" != "xtrue" ]; then
+                exit $ret
+            else
+                return $ret
+            fi
         fi
 
         if [ "x$HSM_TOKEN" != "x" ]; then
             # In the event we have a HSM, we also have to remove it from the
             # HSM token.
 
-            __v certutil "${remove_args[@]}" "-n" "$HSM_TOKEN:$CERT_NICKNAME"
-            certutil "${remove_args[@]}" "-n" "$HSM_TOKEN:$CERT_NICKNAME"
+            __v certutil "${remove_args[@]}" "-n" "$HSM_TOKEN:$nickname"
+            certutil "${remove_args[@]}" "-n" "$HSM_TOKEN:$nickname"
             local ret=$?
+
             if (( ret != 0 )); then
                 __e "certutil returned non-zero result: $ret"
                 __e "Unable to delete certificate!"
-                exit $?
+
+                if [ "x$return_no_exit" != "xtrue" ]; then
+                    exit $ret
+                else
+                    return $ret
+                fi
             fi
         fi
+
+        return 0
+    }
+
+    # Remove the specified key from the NSS DB specified by $1.
+    function _remove_key() {
+        local key_id="$1"
+
+        local remove_args=("-F")
+
+        remove_args+=("-d" "$NSSDB_TYPE$NSSDB")
+        if [ "x$NSSDB_PASSWORD" != "x" ]; then
+            remove_args+=("-f" "$NSSDB_PASSWORD")
+        fi
+
+        __v certutil "${remove_args[@]}" "-k" "$key_id"
+        certutil "${remove_args[@]}" "-k" "$key_id"
+        local ret=$?
+        if (( ret != 0 )); then
+            __e "certutil returned non-zero result: $ret"
+            __e "Unable to delete specified key: $key_id!"
+
+            return $ret
+        fi
+
+        # Since we only remove keys when we're importing a .p12, and we don't
+        # support importing .p12 onto HSMs, we don't need to handle the key
```

Comment:
Maybe because I'm not reviewing code but rather just looking at comments for additional design info, I don't know why you say "we don't support importing .p12 onto HSMs".  Is there a reason for that?

Reply:
If I recall correctly, at the time you said we didn't need to support HSMs and `.p12` as we never import `.p12`s onto HSMs in a CC environment. I can dig up logs on Monday...

I forget exactly what I was concerned about though -- I want to say it was a combination of:
 - Removing keys / certificates when import fails, which we're doing already, but with an HSM requires an extra 2 calls to `certutil -L` and `certutil -K` to track what was on the HSM to begin with, plus potentially doubling the number of deletion calls...
 -  And the overall performance problems:
    - Extra calls to `certutil` with HSM operations above...
    - Neither `certutil` nor `pk12util` nor `openssl` support sorting the certificate chain from root -> leaf (or visa-versa for that matter), so our validation strategy is "try to validate and retry _everything remaining_ (including failed validations) if _anything_ succeeded in the previous step"
        - This is due to setting trust flags after import (as we should, per earlier discussion), but gives us an O(n^2) worst-case performance of validation.
        - Even though `n <= 4` is probably a safe assumption, each HSM operation is **really slow** (up to say, 2-3 minutes), so you're looking at 20 minutes to validate and import a long chain. IMO, that's really, really too long.

When we rewrite this in C or Java and get support for NSS's in-memory NSS DBs (which we can try the import into, do the validation, and safely recover from in the event of validation failure)... This will be a *lot* simpler to support.

Reply:
Actually it's fine.  I remember the conversation now.  I guess I was reading "we don't support importing .p12 onto HSMs" as "we can't support importing .p12 onto HSMs", when what you really meant was "presently we don't have a need to support importing .p12 onto HSMs."  Yes it could get messy dealing with HSM so we don't need to do that unless it is requested.